bridge/github/export.go

```diff
@@ -0,0 +1,654 @@
+	}
+
+	aux := struct {
+		ID     string `json:"id"`
+		NodeID string `json:"node_id"`
+		Color  string `json:"color"`
+	}{}
+
+	data, _ := ioutil.ReadAll(resp.Body)
+	defer resp.Body.Close()
+
+	err = json.Unmarshal(data, &aux)
+	if err != nil {
+		return "", err
+	}
+
+	return aux.NodeID, nil
+}
+
+// create github label using api v3
+func (ge *githubExporter) createGithubLabel(label, labelColor string) (string, error) {
+	url := fmt.Sprintf("%s/repos/%s/%s/labels", githubV3Url, ge.conf[keyOwner], ge.conf[keyProject])
+
+	client := &http.Client{
+		Timeout: defaultTimeout,
+	}
+
+	req, err := http.NewRequest("POST", url, nil)
+	if err != nil {
+		return "", err
+	}
+
+	// need the token for private repositories
+	req.Header.Set("Authorization", fmt.Sprintf("token %s", ge.conf[keyToken]))
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+
+	if resp.StatusCode != http.StatusCreated {
+		return "", fmt.Errorf("error creating label: response status %v", resp.StatusCode)
+	}
+
+	aux := struct {
+		ID     string `json:"id"`
+		NodeID string `json:"node_id"`
+		Color  string `json:"color"`
+	}{}
+
+	data, _ := ioutil.ReadAll(resp.Body)
+	defer resp.Body.Close()
+
+	err = json.Unmarshal(data, &aux)
+	if err != nil {
+		return "", err
+	}
+
+	return aux.NodeID, nil
+}
+
+// randomHexColor return a random hex color code
+func randomHexColor() string {
+	bytes := make([]byte, 6)
+	if _, err := rand.Read(bytes); err != nil {
+		return "fffff"
+	}
+
+	return hex.EncodeToString(bytes)
+}
+
+func (ge *githubExporter) getOrCreateGithubLabelID(repositoryID, label string) (string, error) {
+	// try to get label id
+	labelID, err := ge.getGithubLabelID(label)
+	if err == nil {
+		return labelID, nil
+	}
+
+	// random color
+	//TODO: no random
+	color := randomHexColor()
+
+	// create label and return id
+	labelID, err = ge.createGithubLabel(label, color)
+	if err != nil {
+		return "", err
+	}
+
+	return labelID, nil
+}
+
+func (ge *githubExporter) getLabelsIDs(repositoryID string, labels []bug.Label) ([]githubv4.ID, error) {
+	ids := make([]githubv4.ID, 0, len(labels))
+	var err error
+
+	// check labels ids
+	for _, l := range labels {
+		label := string(l)
+
+		id, ok := ge.cachedLabels[label]
+		if !ok {
+			// try to query label id
+			id, err = ge.getOrCreateGithubLabelID(repositoryID, label)
+			if err != nil {
+				return nil, fmt.Errorf("get or create github label: %v", err)
+			}
+
+			// cache label id
+			ge.cachedLabels[label] = id
+		}
+
+		ids = append(ids, githubv4.ID(id))
+	}
+
+	return ids, nil
+}
+
+// create a github issue and return it ID
+func createGithubIssue(gc *githubv4.Client, repositoryID, title, body string) (string, string, error) {
+	m := &createIssueMutation{}
+	input := &githubv4.CreateIssueInput{
+		RepositoryID: repositoryID,
+		Title:        githubv4.String(title),
+		Body:         (*githubv4.String)(&body),
+	}
+
+	if err := gc.Mutate(context.TODO(), m, input, nil); err != nil {
```

Comment:
You probably need something like:
```
ctx, cancel := context.WithTimeout(ctx, defaultTimeout)
defer cancel()
if err := gc.Mutate(ctx, m, input, nil); err != nil {
```

Reply:
Okkey i'll make the changes to the importer too

Reply:
Should the parent context be created inside the request functions or passed as first parameter ?

Reply:
Good question .... Let's keep it contained here until more is needed.

Reply:
Done !